File: qt-app/src/services/data/DataStore.cpp
#include "DataStore.h"



#include <algorithm>



#include <QDateTime>

#include <QDir>

#include <QFile>

#include <QJsonArray>

#include <QJsonDocument>

#include <QJsonObject>

#include <QRegularExpression>

#include <QStandardPaths>



namespace {

bool atomicWriteJson(const QString& path, const QJsonDocument& document);

QJsonObject holdingToJson(const Holding& holding)

{

    return {

        {QStringLiteral("ticker"), holding.ticker},

        {QStringLiteral("weight"), holding.weight},

        {QStringLiteral("yield"), holding.yield},

        {QStringLiteral("beta"), holding.beta}

    };

}



Holding holdingFromJson(const QJsonObject& object)

{

    Holding holding;

    holding.ticker = object.value(QStringLiteral("ticker")).toString();

    holding.weight = object.value(QStringLiteral("weight")).toDouble();

    holding.yield = object.value(QStringLiteral("yield")).toDouble();

    holding.beta = object.value(QStringLiteral("beta")).toDouble();

    return holding;

}



QJsonObject portfolioToJson(const Portfolio& portfolio)

{

    QJsonArray holdings;

    for (const auto& holding : portfolio.holdings) {

        holdings.append(holdingToJson(holding));

    }



    return {

        {QStringLiteral("id"), portfolio.id},

        {QStringLiteral("name"), portfolio.name},

        {QStringLiteral("targetYield"), portfolio.targetYield},

        {QStringLiteral("aggregateBeta"), portfolio.aggregateBeta},

        {QStringLiteral("achievedYield"), portfolio.achievedYield},

        {QStringLiteral("holdings"), holdings},

        {QStringLiteral("createdAt"), portfolio.createdAt.toString(Qt::ISODate)}

    };

}



Portfolio portfolioFromJson(const QJsonObject& object)

{

    Portfolio portfolio;

    portfolio.id = object.value(QStringLiteral("id")).toString();

    portfolio.name = object.value(QStringLiteral("name")).toString();

    portfolio.targetYield = object.value(QStringLiteral("targetYield")).toDouble();

    portfolio.aggregateBeta = object.value(QStringLiteral("aggregateBeta")).toDouble();

    portfolio.achievedYield = object.value(QStringLiteral("achievedYield")).toDouble();

    portfolio.createdAt = QDateTime::fromString(object.value(QStringLiteral("createdAt")).toString(), Qt::ISODate);



    const auto holdings = object.value(QStringLiteral("holdings")).toArray();

    for (const auto& value : holdings) {

        portfolio.holdings.append(holdingFromJson(value.toObject()));

    }



    return portfolio;

}

bool writePortfolios(const QString& path, const QVector<Portfolio>& portfolios)

{

    QJsonArray items;

    for (const auto& item : portfolios) {

        items.append(portfolioToJson(item));

    }



    return atomicWriteJson(path, QJsonDocument(items));

}



QJsonObject assetToJson(const Asset& asset)

{

    return {

        {QStringLiteral("ticker"), asset.ticker},

        {QStringLiteral("name"), asset.name},

        {QStringLiteral("sector"), asset.sector},

        {QStringLiteral("price"), asset.price},

        {QStringLiteral("dividendYield"), asset.dividendYield},

        {QStringLiteral("beta"), asset.beta},

        {QStringLiteral("marketCap"), asset.marketCap},

        {QStringLiteral("isETF"), asset.isETF},

        {QStringLiteral("lastUpdated"), asset.lastUpdated.toString(Qt::ISODate)}

    };

}



Asset assetFromJson(const QJsonObject& object)

{

    Asset asset;

    asset.ticker = object.value(QStringLiteral("ticker")).toString();

    asset.name = object.value(QStringLiteral("name")).toString();

    asset.sector = object.value(QStringLiteral("sector")).toString();

    asset.price = object.value(QStringLiteral("price")).toDouble();

    asset.dividendYield = object.value(QStringLiteral("dividendYield")).toDouble();

    asset.beta = object.value(QStringLiteral("beta")).toDouble();

    asset.marketCap = object.value(QStringLiteral("marketCap")).toDouble();

    asset.isETF = object.value(QStringLiteral("isETF")).toBool();

    asset.lastUpdated = QDateTime::fromString(object.value(QStringLiteral("lastUpdated")).toString(), Qt::ISODate);

    return asset;

}



QJsonObject sensitivityPointToJson(const SensitivityPoint& point)

{

    QJsonArray holdings;

    for (const auto& holding : point.holdings) {

        holdings.append(holdingToJson(holding));

    }



    return {

        {QStringLiteral("yield"), point.yield},

        {QStringLiteral("beta"), point.beta},

        {QStringLiteral("achievedYield"), point.achievedYield},

        {QStringLiteral("holdings"), holdings}

    };

}



SensitivityPoint sensitivityPointFromJson(const QJsonObject& object)

{

    SensitivityPoint point;

    point.yield = object.value(QStringLiteral("yield")).toDouble();

    point.beta = object.value(QStringLiteral("beta")).toDouble();

    point.achievedYield = object.value(QStringLiteral("achievedYield")).toDouble();



    const auto holdings = object.value(QStringLiteral("holdings")).toArray();

    for (const auto& value : holdings) {

        point.holdings.append(holdingFromJson(value.toObject()));

    }



    return point;

}



QJsonObject sensitivityCurveToJson(const SensitivityCurve& curve)

{

    QJsonArray points;

    for (const auto& point : curve.points) {

        points.append(sensitivityPointToJson(point));

    }



    return {

        {QStringLiteral("minYield"), curve.minYield},

        {QStringLiteral("maxYield"), curve.maxYield},

        {QStringLiteral("step"), curve.step},

        {QStringLiteral("points"), points},

        {QStringLiteral("computedAt"), curve.computedAt.toString(Qt::ISODate)}

    };

}



SensitivityCurve sensitivityCurveFromJson(const QJsonObject& object)

{

    SensitivityCurve curve;

    curve.minYield = object.value(QStringLiteral("minYield")).toDouble();

    curve.maxYield = object.value(QStringLiteral("maxYield")).toDouble();

    curve.step = object.value(QStringLiteral("step")).toDouble();

    curve.computedAt = QDateTime::fromString(object.value(QStringLiteral("computedAt")).toString(), Qt::ISODate);



    const auto points = object.value(QStringLiteral("points")).toArray();

    for (const auto& value : points) {

        curve.points.append(sensitivityPointFromJson(value.toObject()));

    }



    return curve;

}



bool atomicWriteJson(const QString& path, const QJsonDocument& document)

{

    const QString tempPath = path + QStringLiteral(".tmp");

    QFile tempFile(tempPath);

    if (!tempFile.open(QIODevice::WriteOnly | QIODevice::Truncate)) {

        return false;

    }



    tempFile.write(document.toJson(QJsonDocument::Indented));

    tempFile.close();



    QFile::remove(path);

    return QFile::rename(tempPath, path);

}



bool atomicWriteBytes(const QString& path, const QByteArray& payload)

{

    const QString tempPath = path + QStringLiteral(".tmp");

    QFile tempFile(tempPath);

    if (!tempFile.open(QIODevice::WriteOnly | QIODevice::Truncate)) {

        return false;

    }



    tempFile.write(payload);

    tempFile.close();



    QFile::remove(path);

    return QFile::rename(tempPath, path);

}



QString sanitizeBlobKey(QString key)

{

    key.replace(QRegularExpression(QStringLiteral("[^A-Za-z0-9._-]")), QStringLiteral("_"));

    return key;

}

}



bool DataStore::savePortfolio(const Portfolio& portfolio) const

{

    auto portfolios = loadPortfolios();

    const auto it = std::find_if(portfolios.begin(), portfolios.end(), [&](const Portfolio& existing) {

        return existing.id == portfolio.id;

    });



    if (it != portfolios.end()) {

        *it = portfolio;

    } else {

        portfolios.append(portfolio);

    }

    ensureDataDirectory();

    return writePortfolios(portfolioFilePath(), portfolios);

}



QVector<Portfolio> DataStore::loadPortfolios() const

{

    QVector<Portfolio> result;

    QFile file(portfolioFilePath());

    if (!file.exists() || !file.open(QIODevice::ReadOnly)) {

        return result;

    }



    const auto document = QJsonDocument::fromJson(file.readAll());

    for (const auto& value : document.array()) {

        result.append(portfolioFromJson(value.toObject()));

    }



    return result;

}



bool DataStore::deletePortfolio(const QString& id) const

{

    if (id.isEmpty()) {

        return false;

    }



    auto portfolios = loadPortfolios();
    const auto originalSize = portfolios.size();

    portfolios.erase(
        std::remove_if(portfolios.begin(), portfolios.end(), [&](const Portfolio& portfolio) {

            return portfolio.id == id;

        }),
        portfolios.end());

    if (portfolios.size() == originalSize) {

        return false;

    }



    ensureDataDirectory();

    return writePortfolios(portfolioFilePath(), portfolios);

}



bool DataStore::saveAssets(const QVector<Asset>& assets) const

{

    QJsonArray items;

    for (const auto& asset : assets) {

        items.append(assetToJson(asset));

    }



    ensureDataDirectory();

    return atomicWriteJson(assetFilePath(), QJsonDocument(items));

}



QVector<Asset> DataStore::loadAssets() const

{

    QVector<Asset> result;

    QFile file(assetFilePath());

    if (!file.exists() || !file.open(QIODevice::ReadOnly)) {

        return result;

    }



    const auto document = QJsonDocument::fromJson(file.readAll());

    for (const auto& value : document.array()) {

        result.append(assetFromJson(value.toObject()));

    }



    return result;

}



bool DataStore::saveSensitivityCurve(const SensitivityCurve& curve) const

{

    auto curves = loadSensitivityCurves();

    curves.append(curve);



    QJsonArray items;

    for (const auto& item : curves) {

        items.append(sensitivityCurveToJson(item));

    }



    ensureDataDirectory();

    return atomicWriteJson(sensitivityCurveFilePath(), QJsonDocument(items));

}



QVector<SensitivityCurve> DataStore::loadSensitivityCurves() const

{

    QVector<SensitivityCurve> result;

    QFile file(sensitivityCurveFilePath());

    if (!file.exists() || !file.open(QIODevice::ReadOnly)) {

        return result;

    }



    const auto document = QJsonDocument::fromJson(file.readAll());

    for (const auto& value : document.array()) {

        result.append(sensitivityCurveFromJson(value.toObject()));

    }



    return result;

}



bool DataStore::saveBlob(const QString& key, const QByteArray& blob) const

{

    if (key.isEmpty()) {

        return false;

    }



    ensureDataDirectory();

    ensureBlobDirectory();

    return atomicWriteBytes(blobFilePath(key), blob);

}



QByteArray DataStore::loadBlob(const QString& key) const

{

    if (key.isEmpty()) {

        return {};

    }



    QFile file(blobFilePath(key));

    if (!file.exists() || !file.open(QIODevice::ReadOnly)) {

        return {};

    }



    return file.readAll();

}

bool DataStore::saveCachedResponse(const QString& key, const QByteArray& data, int ttlSeconds) const

{

    if (key.isEmpty() || ttlSeconds <= 0 || !saveBlob(key, data)) {

        return false;

    }

    const QJsonObject metadata{
        {QStringLiteral("timestamp"), static_cast<qint64>(QDateTime::currentSecsSinceEpoch())},
        {QStringLiteral("ttl"), ttlSeconds}
    };

    return atomicWriteJson(cacheMetadataFilePath(key), QJsonDocument(metadata));

}

QByteArray DataStore::loadCachedResponse(const QString& key) const

{

    if (!isCacheValid(key)) {
        QFile::remove(blobFilePath(key));
        QFile::remove(cacheMetadataFilePath(key));
        return {};
    }

    return loadBlob(key);

}

bool DataStore::isCacheValid(const QString& key) const

{

    if (key.isEmpty()) {

        return false;

    }

    QFile metadataFile(cacheMetadataFilePath(key));
    if (!metadataFile.exists() || !metadataFile.open(QIODevice::ReadOnly)) {
        return false;
    }

    const QJsonObject metadata = QJsonDocument::fromJson(metadataFile.readAll()).object();
    const qint64 timestamp = metadata.value(QStringLiteral("timestamp")).toVariant().toLongLong();
    const int ttlSeconds = metadata.value(QStringLiteral("ttl")).toInt();
    if (timestamp <= 0 || ttlSeconds <= 0) {
        return false;
    }

    if (!QFile::exists(blobFilePath(key))) {
        return false;
    }

    const qint64 expiresAt = timestamp + ttlSeconds;
    return QDateTime::currentSecsSinceEpoch() <= expiresAt;

}



QString DataStore::dataDirectory() const

{

    return QStandardPaths::writableLocation(QStandardPaths::AppDataLocation);

}



bool DataStore::ensureDataDirectory() const

{

    QDir dir(dataDirectory());

    return dir.exists() || dir.mkpath(QStringLiteral("."));

}



bool DataStore::ensureBlobDirectory() const

{

    QDir dir(dataDirectory());

    return dir.exists(QStringLiteral("blobs")) || dir.mkpath(QStringLiteral("blobs"));

}



QString DataStore::portfolioFilePath() const

{

    return dataDirectory() + QStringLiteral("/portfolios.json");

}



QString DataStore::assetFilePath() const

{

    return dataDirectory() + QStringLiteral("/assets.json");

}



QString DataStore::sensitivityCurveFilePath() const

{

    return dataDirectory() + QStringLiteral("/sensitivity_curves.json");

}



QString DataStore::blobFilePath(const QString& key) const

{

    return dataDirectory() + QStringLiteral("/blobs/") + sanitizeBlobKey(key) + QStringLiteral(".json");

}

QString DataStore::cacheMetadataFilePath(const QString& key) const

{

    return dataDirectory() + QStringLiteral("/blobs/") + sanitizeBlobKey(key) + QStringLiteral(".meta.json");

}

File: src/services/data_store.py
"""Persistent JSON storage for assets, portfolios, and cache."""

import json
import os
from datetime import datetime
from pathlib import Path
from typing import List, Optional

from PySide6.QtCore import QStandardPaths

from ..models import Asset, Holding, Portfolio


def _data_dir() -> Path:
    path = Path(QStandardPaths.writableLocation(QStandardPaths.StandardLocation.AppDataLocation))
    path.mkdir(parents=True, exist_ok=True)
    return path


def _atomic_write(filepath: Path, content: str) -> None:
    tmp = filepath.with_suffix(".tmp")
    tmp.write_text(content, encoding="utf-8")
    if filepath.exists():
        filepath.unlink()
    tmp.rename(filepath)


class DataStore:
    def __init__(self):
        self._dir = _data_dir()

    @property
    def data_directory(self) -> Path:
        return self._dir

    # --- Assets ---
    def load_assets(self) -> List[Asset]:
        path = self._dir / "assets.json"
        if not path.exists():
            return []
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
            return [self._asset_from_dict(d) for d in data]
        except (json.JSONDecodeError, KeyError):
            return []

    def save_assets(self, assets: List[Asset]) -> None:
        path = self._dir / "assets.json"
        data = [self._asset_to_dict(a) for a in assets]
        _atomic_write(path, json.dumps(data, indent=2))

    # --- Portfolios ---
    def load_portfolios(self) -> List[Portfolio]:
        path = self._dir / "portfolios.json"
        if not path.exists():
            return []
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
            return [self._portfolio_from_dict(d) for d in data]
        except (json.JSONDecodeError, KeyError):
            return []

    def save_portfolio(self, portfolio: Portfolio) -> None:
        portfolios = self.load_portfolios()
        existing = next((i for i, p in enumerate(portfolios) if p.id == portfolio.id), None)
        if existing is not None:
            portfolios[existing] = portfolio
        else:
            portfolios.append(portfolio)
        path = self._dir / "portfolios.json"
        data = [self._portfolio_to_dict(p) for p in portfolios]
        _atomic_write(path, json.dumps(data, indent=2))

    def delete_portfolio(self, portfolio_id: str) -> bool:
        portfolios = self.load_portfolios()
        filtered = [p for p in portfolios if p.id != portfolio_id]
        if len(filtered) == len(portfolios):
            return False
        path = self._dir / "portfolios.json"
        _atomic_write(path, json.dumps([self._portfolio_to_dict(p) for p in filtered], indent=2))
        return True

    # --- Serialization helpers ---
    @staticmethod
    def _asset_to_dict(asset: Asset) -> dict:
        return {
            "ticker": asset.ticker,
            "name": asset.name,
            "sector": asset.sector,
            "price": asset.price,
            "dividendYield": asset.dividend_yield,
            "beta": asset.beta,
            "marketCap": asset.market_cap,
            "isETF": asset.is_etf,
            "lastUpdated": asset.last_updated.isoformat() if asset.last_updated else None,
        }

    @staticmethod
    def _asset_from_dict(d: dict) -> Asset:
        last_updated = None
        if d.get("lastUpdated"):
            try:
                last_updated = datetime.fromisoformat(d["lastUpdated"])
            except ValueError:
                last_updated = datetime.utcnow()
        return Asset(
            ticker=d.get("ticker", ""),
            name=d.get("name", ""),
            sector=d.get("sector", ""),
            price=d.get("price", 0.0),
            dividend_yield=d.get("dividendYield", 0.0),
            beta=d.get("beta", 0.0),
            market_cap=d.get("marketCap", 0.0),
            is_etf=d.get("isETF", False),
            last_updated=last_updated or datetime.utcnow(),
        )

    @staticmethod
    def _holding_to_dict(h: Holding) -> dict:
        return {"ticker": h.ticker, "weight": h.weight, "yield": h.dividend_yield, "beta": h.beta}

    @staticmethod
    def _holding_from_dict(d: dict) -> Holding:
        return Holding(
            ticker=d.get("ticker", ""),
            weight=d.get("weight", 0.0),
            dividend_yield=d.get("yield", 0.0),
            beta=d.get("beta", 0.0),
        )

    @staticmethod
    def _portfolio_to_dict(p: Portfolio) -> dict:
        return {
            "id": p.id,
            "name": p.name,
            "targetYield": p.target_yield,
            "achievedYield": p.achieved_yield,
            "aggregateBeta": p.aggregate_beta,
            "holdings": [DataStore._holding_to_dict(h) for h in p.holdings],
            "createdAt": p.created_at.isoformat() if p.created_at else None,
        }

    @staticmethod
    def _portfolio_from_dict(d: dict) -> Portfolio:
        created_at = None
        if d.get("createdAt"):
            try:
                created_at = datetime.fromisoformat(d["createdAt"])
            except ValueError:
                created_at = datetime.utcnow()
        holdings = [DataStore._holding_from_dict(h) for h in d.get("holdings", [])]
        return Portfolio(
            id=d.get("id", ""),
            name=d.get("name", ""),
            target_yield=d.get("targetYield", 0.0),
            achieved_yield=d.get("achievedYield", 0.0),
            aggregate_beta=d.get("aggregateBeta", 0.0),
            holdings=holdings,
            created_at=created_at or datetime.utcnow(),
        )
